File: mid/Hashing/347. Top K Frequent Elements.cpp
// https://leetcode.com/problems/top-k-frequent-elements/solutions/4580564/5-1-approach-1-o-n-log-n-python-c-step-by-step-explanation/
// Time complexity : O(n log n)
// Space complexity : O(n)
// Hash Map

class Solution {
    static bool cmp(pair<int, int>& a, pair<int, int>& b) {
        return a.second > b.second;
    }
public:
    vector<int> topKFrequent(vector<int>& nums, int k) {
        // Step 1: Use an unordered_map (mp) to count the frequency of each element in nums.
        unordered_map<int, int> mp;
        for (auto x : nums) {
            mp[x]++;
        }

        // Step 2: Create a vector of pairs (v) to store the elements and their frequencies.
        vector<pair<int, int>> v;
        for (auto x : mp) {
            v.push_back(pair{x.first, x.second});
        }

        // Step 3: Sort the vector based on frequencies in descending order.
        sort(v.begin(), v.end(), cmp);

        // Step 4: Create a vector (ans) to store the top k frequent elements.
        vector<int> ans;
        for (int i = 0; i < k; i++) {
            auto it = v.begin() + i;
            ans.push_back(it->first);
        }
        return ans;
    }
};
        

File: mid/Hashing/347. Top K Frequent Elements.py
# https://leetcode.com/problems/top-k-frequent-elements/solutions/1928198/python-simple-python-solution-using-dictionary-hashmap/
# Time Complexity : O(n * log(n)) (n + nlogn(sorted())= nlogn)
# Space Complexity : O(n)
# Beats 90%

class Solution:
    def topKFrequent(self, nums: List[int], k: int) -> List[int]:
        frequency = {}
        for num in nums:
            if num not in frequency:
                frequency[num] = 1
            else:
                frequency[num] += 1

        frequency = dict(sorted(frequency.items(), key=lambda x: x[1], reverse=True))
        result = list(frequency.keys())[:k]
        
        return result
